ironfish/src/network/peers/addressManager.ts

```diff
@@ -2,24 +2,149 @@
+   * filter out peer addresses. It attempts to find a previously-connected
+   * peer address that is not part of an active connection. If there are none,
+   * then it attempts to find a previously-known peer address.
+   */
+  getRandomDisconnectedPeerAddress(peers: Peer[]): PeerAddress | null {
+    if (
+      this.possiblePeerAddresses.length === 0 &&
+      this.priorConnectedPeerAddresses.length === 0
+    ) {
+      return null
+    }
+
+    const currentPeerAddresses = new Set(
+      //eslint-disable-next-line @typescript-eslint/no-non-null-assertion
+      peers.filter((peer) => peer.state.identity !== null).map((peer) => peer.state.identity!),
+    )
+
+    const priorConnectedAddressSet = new Set([...this.priorConnectedPeerAddresses])
+
+    const disconnectedPriorAddresses = this.filterConnectedAddresses(
+      priorConnectedAddressSet,
+      currentPeerAddresses,
+    )
+    if (disconnectedPriorAddresses.length) {
+      return ArrayUtils.sampleOrThrow(disconnectedPriorAddresses)
+    } else {
+      const possibleAddressSet = new Set([...this.possiblePeerAddresses])
+      const disconnectedPossibleAddresses = this.filterConnectedAddresses(
+        possibleAddressSet,
+        currentPeerAddresses,
+      )
+      if (disconnectedPossibleAddresses.length) {
+        return ArrayUtils.sampleOrThrow(disconnectedPossibleAddresses)
+      } else {
+        return null
+      }
+    }
+  }
+
+  private filterConnectedAddresses(
+    addressSet: Set<Readonly<PeerAddress>>,
+    connectedPeerAdresses: Set<string>,
+  ): PeerAddress[] {
+    const disconnectedAddresses = [...addressSet].filter(
+      (address) => address.identity !== null && !connectedPeerAdresses.has(address.identity),
+    )
+
+    return disconnectedAddresses
+  }
+
+  /**
+   * Removes address associated with a peer from address stores
+   */
+  removePeerAddress(peer: Peer): void {
+    const filteredPossibles = this.possiblePeerAddresses.filter(
+      (possible) => possible.identity !== peer.state.identity,
+    )
+    const filteredPriorConnected = this.priorConnectedPeerAddresses.filter(
+      (prior) => prior.identity !== peer.state.identity,
+    )
+
+    this.hostsStore.set('possiblePeers', filteredPossibles)
+    this.hostsStore.set('priorPeers', filteredPriorConnected)
+  }
+
+  /**
+   * Persist all currently connected peers and unused peer addresses to disk
+   */
+  async save(peers: Peer[]): Promise<void> {
+    const inUsePeerAddresses = peers
+      .filter(
+        (peer) =>
+          peer.state.type === 'CONNECTED' &&
+          !peer.getConnectionRetry(ConnectionType.WebSocket, ConnectionDirection.Outbound)
+            ?.willNeverRetryConnecting,
+      )
+      .map((peer) => ({
+        address: peer.address,
+        port: peer.port,
+        identity: peer.state.identity ?? null,
+        name: peer.name ?? null,
+      }))
+    this.hostsStore.set('priorPeers', [...inUsePeerAddresses])
```

Comment:
The value of `inUsePeerAddresses` is not exactly what we want for `priorPeers` -- I think we'd want all peers whom we've successfully established an outbound WebSocket connection to at least once in the past, but may or may not be currently connected to.

This is probably a non-trivial change, so now that I think about it, I'm not sure it's worth doing in this PR, but maybe a comment mentioning that would be nice.